### ISSUE TITLE
Social login with google

### ISSUE THREAD

**Issue by MattBlack85** [Tivix/django-rest-auth#262]
Hi, I want to provide social login via facebook/google for my app, While I got FB to work I cannot figure out what the `code` should be when using google as auth provider. Could please anybody elaborate on that?

Also I think we should document that for facebook `code` is the same of `App Token`, it took me some time to figure it out

**Comment by gbrennon** [comment]
I'm having the same issue....

**Comment by MattBlack85** [comment]
@gbrennon I was able to use also google login in the end, I'm putting together a small tutorial but for now:
- Add to your google loign view `client_class = OAuth2Client`
- Add to the same view a callback url (it can be a fake one)
- now you are able to log your users in only with `code`

If you want to test is you can use google you can use https://developers.google.com/oauthplayground/

- choose Google OAuth API v2
- choose some of the options
- click authorize API
- you are redirected to google profiles choose
- allow
Now on the left side of the page you can see an `authorization code`, that's the `code` required by rest-auth, have fun!

The playgorund simulate well a mobile app asking for permissions, then you can create a real app in the developer's console, put the required code into the mobile/web app and send the `Authorization code` from google to your API.

Hope this helps!

**Comment by veris-neerajdhiman** [comment]
@MattBlack85 Hi,

I used playground simulation to get `authorization code` but when I use that code in `res-auth` API. I am getting below error. 

```json
{
    "non_field_errors": [
        "Incorrect value"
    ]
}
```

Any idea what's the issue ?

**Comment by MattBlack85** [comment]
@veris-neerajdhiman it seems like your code is not valid, weird... Did you set up the social app within the database?

**Comment by veris-neerajdhiman** [comment]
@MattBlack85 yes in db, but the access token I am getting in exchanged of authorization is working well. 

seems like a bug in rest-auth, fixed in this pull request https://github.com/Tivix/django-rest-auth/pull/283

**Comment by olivierdalang** [comment]
Hi,

I'm also stuck on this, and I'm not sure whether it's a bug or I'm not understanding how it works.
Basically, my django-rest-auth endpoints accepts two parameters : `code` and `auth_token`.
Using Ionic2's Google plugin, I can log my client app to google, and get this (I hid sensitive values): 
```
{
    "email": "***",
    "idToken": "************************************************************************************************",
    "serverAuthCode": "4/******************",
    "userId": "***",
    "displayName": "***",
    "familyName": "***",
    "givenName": "***",
    "imageUrl": "https://lh3.googleusercontent.com/***/photo.jpg"
}
```

Now I tried to enter all combination between my api's `code` and `auth_token` and my client's `idToken` and `serverAuth` codes, and am always getting 
```
{
    "non_field_errors": [
        "Incorrect value"
    ]
}
```

It is possible that I misconfigured the Google API keys ?
I have one key for the client (an Android key, that seems to work since I'm getting values back) and a different key for my server (an OAuth2 key).

I think the doc should be a little bit more explicit about how this works !

Thanks for the awesome work !

**Comment by teolemon** [comment]
@MattBlack85 Thank for the steps. Now in my case, that's not an app requesting, but Actions on Google (for Google Assistant integration for your chatbot, with Google account linking). 
In that case, it fails when I test it with https://gala-demo.appspot.com/app. (Assertion response must contain one of 'error' or 'access_token')
You didn't mean to put Authorization code you get in the Oauth playground into the view as ```code="XXXsd…"``` , right ?

**Comment by NDevox** [comment]
@MattBlack85 did you get anywhere with that tutorial?

I've set up the API credentials on google giving the authorised JavaScript origins and authorized redirect URI as `http://localhost:8000` and `http://localhost:8000/accounts/google/login/callback/`respectively. I've added a social application with the client id and secret key with Google as the provider.

I go to the oauth playground tick all the Google OAuth2 scopes and get the authorization code.

I send a POST to `localhost:8000/rest-auth/google/` with the authroization code in a form field called `code` but the response I get from google is a 401 unauthorized with the content `{"error" : "unauthorized_client"}` which means nothing to me.

Have I missed anything?

**Comment by umarmughal824** [comment]
Did anyone found the solution of this problem?

"non_field_errors": [
        "Incorrect value"
    ]

with correct access_token posting.

**Comment by MattBlack85** [comment]
I had this working until our app was shut down. I'll try to dig through the code I made and see if I can replicate the login

**Comment by umarmughal824** [comment]
Hey guys, I have fixed it by customization. I have posted in on StackOverflow as well.

Here is the link of my stackoverflow answer. 

[see the answer of this issue](https://stackoverflow.com/questions/42454082/django-rest-auth-with-allauth/48182227#48182227)
